### ISSUE TITLE
Scripts for spot instance multiplayer gaming.

### ISSUE THREAD

**Issue by siliconhippy** [keithvassallomt/parsec-aws-automation#5]
Hi Keith !

I had commented on your excellent youtube some  weeks ago, and am wondering about the multiplayer scenario:

Parsec already registers player accounts that are authorized to access the admin player account.

So how would you construct an automated stop- start script that says: 

Shut down if there are no players present. But start up if there is at least one player present ( unless already started.)

How would this work then with the Spot instance script? If I don't user this Spot script, will the instance start in regular on demand more?

I guess the trick is to determine a trigger event that says a player(s)has logged in, or there is no one logged in ( including admin.)

There are some Windows game control panel packages I read about online ( tcadmin, and Cube Coders AMP) for multiplayer Server gaming, but I don't know how they'll be useful here with one game copy install.

Looking forward to a multiplayer youtube as well !

**Comment by keithvassallomt** [comment]
Hey,

That’s an interesting idea. The biggest challenge would be if someone other than the machine owner wants to start gaming. An instance would need to be spun up for that user, which requires AWS credentials. This means that your friends would have just as much access to the server as you do :/

The alternative I see is to create an API endpoint with some simple web interface. When a friend wants to game, they hit the endpoint which brings up the server.

As for shutting down after all players log out, I’m sure there’s some way to do it via Lambda, but that requires further research.

Honestly I’m not much into multiplayer or sharing the instance, although the latter would be a great way to reduce cost. 

I’ll be updating this guide in the coming weeks since some AWS consoles have changed. Maybe I can look into this, but depends on time. I’m also interested in setting this up with cloud formation, so users could do it with just one click.

Keith

**Comment by siliconhippy** [comment]
Keith,

Thanks for the quick Sunday response 🐱!

Unluckily I am not a tech guy, just a hobbyist with a product manager background.

1. As long as the gpu instance setup is capable, why do non admin players need an instance each for themselves?

I thought the whole cloud gaming idea is that the admin player sets up the gpu instance, purchases/ acquires the game copy to be installed in the AWS instance via Steam ( or Epic etc), and then additional players without admin creds could be invited ( via Parsec or Steam Remote Play accounts) to play on the running instance controlled by admin.

Am I missing something? How many players do you think an AWS g4 instance could accommodate for say GTA5/ Dota type game play?

2. The API endpoint with simple web interface makes sense- there must be some AWS feature already.

Maybe this API connects to a separate minimal AWS instance (always running !) that is managing the GPU game instance, because this game instance would need to be started once incoming player requests come in, and then stopped some poll time after  no players ( including admin) are present.

Of course multiple GPU game instances could also be managed by a single manager instance in a more complex setup. Same thing as saying that this manager instance is like the tcadmin or AMP game control panel but can include both the game control panel and the instance start-stop scheduler.

3. Your single player scripts seem to start/ stop instance based on admin player logging on/off.

Since other players are non admin, a Lambda trigger would need to be generated that defines  player log on/off events, perhaps based on some parameter like incoming connect request, etc.

Maybe Parsec/ Steam client IDs  (already available to instance via invite acceptance) can be parsed from the incoming request, much as a regular server handling user accounts ?

**Comment by keithvassallomt** [comment]
> I thought the whole cloud gaming idea is that the admin player sets up the gpu instance, purchases/ acquires the game copy to be installed in the AWS instance via Steam ( or Epic etc), and then additional players without admin creds could be invited ( via Parsec or Steam Remote Play accounts) to play on the running instance controlled by admin.

The problem with that is the instance would need to keep running 24/7 to be able to allow players to connect when they wish. See, with my scripts, the user spins up an instance (since they are admin), play, and then the instance is automatically terminated and a snapshot created when they shutdown. Since a guest would not be admin, they would not have permission to spin up an instance themselves, meaning the instance would need to keep running indefinitely, which would be *VERY* expensive.

> How many players do you think an AWS g4 instance could accommodate for say GTA5/ Dota type game play?

Depends. A g4dn.xlarge instance could handle quite a few players, playing together on multiplayer.

> Maybe this API connects to a separate minimal AWS instance (always running !) that is managing the GPU game instance, because this game instance would need to be started once incoming player requests come in, and then stopped some poll time after no players ( including admin) are present.

Pretty good idea :) Although this assumes you want to allow multiplayer gaming when the admin is not also playing - which could be a good use case.

> 1. Your single player scripts seem to start/ stop instance based on admin player logging on/off.

Actually, no. The script latches on to the event of the instance terminating. This means the player shuts down the instance when they are done playing, and that's when the script runs. There's no monitoring of logging in or out.

I'm quite sure this *could* in theory be done. However, it would require quite a lot of research and code, and I still think it would not be financially viable compared to a commercially available cloud gaming service.

**Comment by siliconhippy** [comment]
" Pretty good idea :) Although this assumes you want to allow multiplayer gaming when the admin is not also playing - which could be a good use case. "

Yes 😂 I believe in this use case- a bunch of friends playing together in a single GPU instance, or on their own as single player, without necessarily the admin player  always present.

So here the instance manager instance detects logins/outs, pretty much like a game control panel ( AMP, tcadmin, cpanel for general websites ) would, then has automated admin permissions to stop or start the GPU game instance.

The trick then is to use this login detect feature or just integrate instance manager script with above game control panels. Such login might trigger some defined event like incoming request to instance manager or some load condition; the manager then stops/ starts game instance and connects incoming login requests to the game instance. Logout might trigger some other event or change of load.

Basically all players need login to the instance manager server, in addition to being accepted by Parsec/ Steam Remote Play accounts, the latter happening automatically once setup. Of course only the admin player has admin rights ( changing configuration, whitelists, allowing auto operation of game instance without admin presence.)

Maybe this is not as complex as commercial game platforms like Amazon AWS Gamelift ( although I have just read their intro doc and have no clue what features they actually offer !)

Btw, commercial platforms should be more concerned with reserved instances and autoscaling, aka instances management at scale.

In any case thanks for the discussion. I learned a lot ...
